src/libostree/ostree-repo-checkout.c

```diff
@@ -154,129 +154,57 @@ write_regular_file_content (OstreeRepo            *self,
-      guint32 file_mode;
-
-      file_mode = g_file_info_get_attribute_uint32 (file_info, "unix::mode");
-      /* Don't make setuid files on checkout when we're doing --user */
-      if (options->mode == OSTREE_REPO_CHECKOUT_MODE_USER)
-        file_mode &= ~(S_ISUID|S_ISGID);
-
-      do
-        fd = openat (destination_dfd, destination_name, O_WRONLY | O_CREAT | O_EXCL, file_mode);
-      while (G_UNLIKELY (fd == -1 && errno == EINTR));
-      if (fd == -1)
+      if (symlinkat (g_file_info_get_symlink_target (file_info),
+                     destination_dfd, destination_name) < 0)
         {
-          if (errno == EEXIST && options->overwrite_mode == OSTREE_REPO_CHECKOUT_OVERWRITE_ADD_FILES)
+          /* Handle union/add behaviors if we get EEXIST */
+          if (errno == EEXIST && union_mode)
             {
-              /* Note early return */
-              return TRUE;
+              /* Unioning?  Let's unlink and try again */
+              (void) unlinkat (destination_dfd, destination_name, 0);
+              if (symlinkat (g_file_info_get_symlink_target (file_info),
+                             destination_dfd, destination_name) < 0)
+                return glnx_throw_errno (error);
             }
+          else if (errno == EEXIST && add_mode)
+            ;
           else
             return glnx_throw_errno (error);
         }
-      temp_out = g_unix_output_stream_new (fd, TRUE);
-      fd = -1; /* Transfer ownership */
-
-      if (!write_regular_file_content (self, options, temp_out, file_info, xattrs, input,
-                                       cancellable, error))
-        return FALSE;
-    }
-  else
-    g_assert_not_reached ();
-
-  return TRUE;
-}
-
-/*
- * This function creates a file under a temporary name, then rename()s
- * it into place.  This implements union-like behavior.
- */
-static gboolean
-checkout_file_unioning_from_input_at (OstreeRepo     *repo,
-                                      OstreeRepoCheckoutAtOptions  *options,
-                                      GFileInfo      *file_info,
-                                      GVariant       *xattrs,
-                                      GInputStream   *input,
-                                      int             destination_dfd,
-                                      const char     *destination_name,
-                                      GCancellable   *cancellable,
-                                      GError        **error)
-{
-  g_autofree char *temp_filename = NULL;
-
-  if (g_file_info_get_file_type (file_info) == G_FILE_TYPE_SYMBOLIC_LINK)
-    {
-      if (!_ostree_make_temporary_symlink_at (destination_dfd,
-                                              g_file_info_get_symlink_target (file_info),
-                                              &temp_filename,
-                                              cancellable, error))
-        return FALSE;
-
+      /* Process any xattrs now that we made the link */
       if (xattrs)
```

Comment:
We still need the `chown` bit too, right? Otherwise, I see `ostree diff` immediately reporting something after a fresh checkout:

```
# ostree ls symlink-chown-test
d00755 0 0      0 /
l00777 1000 1000      0 /mylink -> mytarget
# mkdir tmp/copy-checkout && mount --bind tmp/copy-checkout tmp/copy-checkout
# ostree checkout symlink-chown-test tmp/copy-checkout/checkout
# ostree diff symlink-chown-test ./tmp/copy-checkout/checkout
M    /mylink
```